### PR TITLE
upgrade to Spark 2.0

### PR DESCRIPTION
**Pull request by skaarthik**

### CODE REVIEW

File: scala/pom.xml [hunk 1]
```diff
@@ -79,33 +79,33 @@
...
-          <artifactId>spark-streaming_2.10</artifactId>
+          <artifactId>spark-streaming_2.11</artifactId>
           <version>${spark.version}</version>
           <!--the following is placeholder for building uber package. Please keep as-is-->
 	      <!--<scope>provided</scope>-->
       </dependency>
       <dependency>
           <groupId>org.apache.spark</groupId>
-          <artifactId>spark-streaming-kafka_2.10</artifactId>
+          <artifactId>spark-streaming-kafka-0-8_2.11</artifactId>
```
Comment (qualiu) [inline scala/pom.xml]: Is this a special version other than spark-streaming-kafka_2.11 ?

**Comment by qualiu** [comment]
Need update scala\perf\pom.xml
image

**Comment by qualiu** [comment]
And I saw some packages kept 2.10 not upgraded to 2.11, intended to keep them ?
image

File: scala/pom.xml (hunk 1 above)
Comment (skaarthik) [inline scala/pom.xml]: Two different versions of this library exist to support two different versions of Kafka

**Comment by skaarthik** [comment]
@qualiu - will update the pom file in perf suite. The other references are just string values in test suite. No need to fix them.

**Comment by qualiu** [comment]
One "preemptive issue" occasionally found and derived : **Double-Quoting spark-submit arguments**

#  Detail as following
When I use  [test.bat] (https://github.com/qualiu/testMobius/blob/master/csharp/testArgsQuotes/test.bat) to test this upgrading : (**3 arguments** as following 1 command line)
`testMobius\csharp\testArgsQuotes\test.bat Pi* d:\tmp "jdbc:mysql://localhost:3306/lzdb?user=guest&password=abc123"
`
I got following error message with 2 errors:
```
d:\msgit\skaarthikMobius\testMobius\csharp\testArgsQuotes\bin\Release>cmd /V /E /C "d:\msgit\skaarthikMobius\build\tools\spark-2.0.0-bin-hadoop2.6\bin\spark-submit2.cmd" --executor-cores 8 --driver-cores 8 --executor-memory 1G --driver-memory 1G --name skaarthikMobius-sparkupgradev2-db85fe3-testArgsQuotes.exe__-ec-8_-dc-8_-em-1G_-dm-1G --class org.apache.spark.deploy.csharp.CSharpRunner D:\msgit\skaarthikMobius\build\runtime\lib\spark-clr_2.11-2.0.000-SNAPSHOT.jar d:\msgit\skaarthikMobius\testMobius\csharp\testArgsQuotes\bin\Release d:\msgit\skaarthikMobius\testMobius\csharp\testArgsQuotes\bin\Release\testArgsQuotes.exe  Pi* d:\tmp "jdbc:mysql://localhost:3306/lzdb?user=guest&password=abc123"
The input line is too long.
'password' is not recognized as an internal or external command,
operable program or batch file
```
* **The input line is too long** : caused by extra double quotes introduced in Spark 2.0
* **'password' is not recognized**  : last argument cut off

# Root cause & Solution-in-Spark + Todo-in-Mobius

## Rough fix in Spark 2.0 [spark-submit.cmd](https://github.com/apache/spark/blob/master/bin/spark-submit.cmd)
### The fix was kind of rude : 
https://github.com/apache/spark/blob/master/bin/spark-submit.cmd#L23
`cmd /V /E /C "%~dp0spark-submit2.cmd" %*`
### A thoughtful fix should check if need quotes.
Or else cannot work around the added double-quotes , and will get "The input line is too long" error.
A fix could be like following :
```
echo "%~dp0spark-submit2.cmd"|findstr /C:" "
if %ERRORLEVEL% EQU 0 (
    cmd /V /E /C "%~dp0spark-submit2.cmd" %*
) else (
    cmd /V /E /C %~dp0spark-submit2.cmd %*
)
```

## To do in Mobius : Support path contains space
I realized another issue after I corrected the [spark-submit.cmd]((https://github.com/apache/spark/blob/master/bin/spark-submit.cmd) :
Does this fix meaningful to Mobius :
Then I test it : just copy {Mobius} to {Mobius - Copy} that contains white space in the directory name.
And a big issue about path-contains-space comes up:
Currently Mobius does not support path contains space.
Just run {Mobius}\build\localmode\RunSamples.cmd will get error. 

# Additional 
This problem has been discovered in last month, but we use the test case in https://github.com/Microsoft/Mobius/tree/master/csharp/Samples
%SPARKCLR_HOME%\scripts\sparkclr-submit.cmd --exe SparkCLRSamples.exe %CD% --torun Pi* --data d:\tmp\skdata%SPARKCLR_HOME%\scripts\sparkclr-submit.cmd --exe SparkCLRSamples.exe %CD% --torun Pi* --data d:\tmp\skdata
The key arguments were : Pi*  d:\tmp\skdata
The sample test is inadequate, so got a false illusion that fixed around Spark-2.0.
So this time I used {testMobius} and added "jdbc:mysql://localhost:3306/lzdb?user=guest&password=abc123".

## An easy way to debug batch (.cmd , .bat) 
* [turn-on-bat-debug-in-dir.bat](https://github.com/qualiu/testMobius/blob/master/tools/bat/turn-on-bat-debug-in-dir.bat)  {Directory-To-do}  -r -R
* [turn-off-bat-debug-in-dir.bat](https://github.com/qualiu/testMobius/blob/master/tools/bat/turn-off-bat-debug-in-dir.bat) {Directory-To-do}  -r -R

**Comment by skaarthik** [comment]
@qualiu can you file a separate issue for the double quote issue?

**Comment by qualiu** [comment]
@skaarthik : Just come back from lunch, I've fired a new issue :https://github.com/Microsoft/Mobius/issues/540

**Comment by qualiu** [comment]
@skaarthik : Another issue , how about submitting a pull request to Spark 2.0 for the case of arguments contains special characters (**mentioned above**)  ?

I found cannot work around the Spark 2.0 with following 1 command line test :
[testMobius\csharp\testArgsQuotes\test.bat](https://github.com/qualiu/testMobius/blob/master/csharp/testArgsQuotes/test.bat)  "jdbc:mysql://localhost:3306/lzdb?user=guest&password=abc123" `

**Comment by qualiu** [comment]
@tawan0109  : The following is what you want to know in detail
# To be more clear for the issue in **Spark-2.0**
The test environment and final **spark-submit.cmd** with arguments as following:
```
HADOOP_HOME=d:\msgit\skaarthikMobius\build\tools\winutils
SPARKCLR_HOME=d:\msgit\skaarthikMobius\build\runtime
SPARK_HOME=d:\msgit\skaarthikMobius\build\tools\spark-2.0.0-bin-hadoop2.6
cmd /V /E /C "d:\msgit\skaarthikMobius\build\tools\spark-2.0.0-bin-hadoop2.6\bin\spark-submit2.cmd" --executor-cores 8 --driver-cores 8 --executor-memory 1G --driver-memory 1G --name skaarthikMobius-sparkupgradev2-db85fe3-testArgsQuotes.exe__-ec-8_-dc-8_-em-1G_-dm-1G --class org.apache.spark.deploy.csharp.CSharpRunner D:\msgit\skaarthikMobius\build\runtime\lib\spark-clr_2.11-2.0.000-SNAPSHOT.jar d:\msgit\skaarthikMobius\testMobius\csharp\testArgsQuotes\bin\Release d:\msgit\skaarthikMobius\testMobius\csharp\testArgsQuotes\bin\Release\testArgsQuotes.exe  Pi* d:\tmp "jdbc:mysql://localhost:3306/lzdb?user=guest&password=abc123"
```
In fact, you can just run a simple command :
```
d:\msgit\skaarthikMobius>cmd /V /E /C "d:\msgit\skaarthikMobius\build\tools\spark-2.0.0-bin-hadoop2.6\bin\spark-submit2.cmd" "jdbc:mysql://localhost:3306/lzdb?user=guest&password=abc123"
```
You'll get the same error as the first full command (which is also previously mentioned):
```
'd:\msgit\skaarthikMobius\build\tools\spark-2.0.0-bin-hadoop2.6\bin\spark-submit2.cmd" "jdbc:mysql://localhost:3306/lzdb?user' is not recognized as an internal or external command,
operable program or batch file.
'password' is not recognized as an internal or external command,
operable program or batch file.
```

### Nested cmd call  cannot accept 2 more quoted arguments
Another way to make it work in https://github.com/apache/spark/blob/master/bin/spark-submit.cmd#L23
Change `cmd /V /E /C "%~dp0spark-submit2.cmd" %*` to 
`"%~dp0spark-submit2.cmd" %*` or `call "%~dp0spark-submit2.cmd" %*`

I checked the usage for the 3 arguments:
* /V : Enable delayed environment variable expansion(such as **!VarName!** , normal is **%VarName%**)
* /E : Enable command extensions
* /C : Carries out the command specified by string and then terminates

But there's no delay expansion usage  like **!VarName!**  in batch file in the whole directory : https://github.com/apache/spark/tree/master/bin
And all of the above settings more generally set in the individual batch files themselves, such as : 
"SetLocal EnableDelayedExpansion" or "SetLocal EnableExtensions EnableDelayedExpansion"

**Comment by qualiu** [comment]
I plan to report this kind of issue case to Spark 2.0, how about your opinions ?

**Comment by qualiu** [comment]
@skaarthik : I've push my commit to my own repository https://github.com/qualiu/spark/tree/submit-cmd-all before sending pull request to community Spark.

**Pull request closed without merging**